### ISSUE TITLE
Token directory should not default to anything in the installed package

### ISSUE THREAD

**Issue by con-f-use** [tybug/ossapi#82]
Currently you have:

```
        self.token_directory = (
            Path(token_directory) if token_directory else Path(__file__).parent
        )
```

in https://github.com/tybug/ossapi/blob/387cb537c185f30f95ef0f0a2cf06511f960e282/ossapi/ossapiv2.py#L492 which is a bad default e.g.  [here](https://github.com/tybug/ossapi/blob/387cb537c185f30f95ef0f0a2cf06511f960e282/ossapi/ossapiv2.py#L683):

```python
    def _save_token(self, token):
        """
        Saves the token to this OssapiV2's associated token file.
        """
        self.log.info(f"saving token to {self.token_file}")
        with open(self.token_file, "wb+") as f:
            pickle.dump(token, f)
```

because this is against python packaging best practices: one should not expect to be the location of the installed package to be writeable or even or even an actual directory/file on the filesystem. Leading to errors like this on some systems:

```
OSError: [Errno 30] Read-only file system: '/home/confus/devel/project/venv/lib/python3.11/site-packages/ossapi/e1e06....708f7.pickle'
```

Ossapi should probably use something like [platformdirs](https://pypi.org/project/platformdirs/) or [`importlib.resources`](https://docs.python.org/3/library/importlib.resources.html), which uses sane defaults and the [XDG spec](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) to determine this default location.

**Comment by tybug** [comment]
Thanks for the report! I had mostly intentionally chosen the installed package location to write these tokens as I didn't want clashes between separate installs of ossapi in different projects. I'll have to think more about the tradeoffs here.

I believe `importlib.resources` is a read-only interface and doesn't let you write resources at run time, and `platformdirs` appdata / etc has the problem of colliding between installs.

FWIW, I'm very aware writing to the install location at runtime is not good practice 😄. However, the largest audience for this package is new programmers messing around with an api for the first time, or using it for low-friction scripts, and I want authentication caching to "just work" for them. I provide the `token_directory` parameter to control this behavior for users which need it.

**Comment by con-f-use** [comment]
> However, the largest audience for this package is new programmers messing around with an api for the first time, or using it for low-friction scripts, and I want authentication caching to "just work" for them.

What prevents you from using platformdirs then? You could even last-resort the package-location if something goes wrong. In snaps or appimages, or on NixOS or Guix your approach will not "just work" :tm:

**Comment by tybug** [comment]
Like I mentioned, I don't want to use platformdirs in order to avoid one install affecting the other, and it needs to persist across invocations, so a temp dir is out.

I'll probably improve the error message here to mention `token_directory` and do a better job of checking ahead of time for read-only filesystems. I don't think changing semantics here is worthwhile.

**Comment by con-f-use** [comment]
Okay, I'm gonna back-off. A final suggestion: if you want something unique per install, you could generate a hash of the `__file__` path and use that to create a directory in proper, writeable location.

**Comment by tybug** [comment]
Yeah, `hash(__file__)` -> `appdata` isn't a bad idea. I'll consider it.
